services/ftp/ftp_service.py

```diff
@@ -0,0 +1,173 @@
+        })
+
+    def on_file_sent(self, file):
+        self.__send_alert(USER_DOWNLOADED_FILE_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(file),
+        })
+
+    def on_file_received(self, file):
+        self.__send_alert(USER_UPLOADED_FILE_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(file),
+        })
+
+    # On the next section we override actual handlers instead of callbacks for more alerts
+    def ftp_LIST(self, path):
+        self.__send_alert(USER_LISTED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_LIST(self, path)
+    
+    def ftp_NLST(self, path):
+        self.__send_alert(USER_LISTED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_NLST(self, path)
+    
+    def ftp_MLST(self, path):
+        self.__send_alert(USER_LISTED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_MLST(self, path)
+    
+    def ftp_CWD(self, path):
+        self.__send_alert(USER_NAVIGATED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_CWD(self, path)
+    
+    def ftp_MKD(self, path):
+        self.__send_alert(USER_CREATED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_MKD(self, path)
+    
+    def ftp_RMD(self, path):
+        self.__send_alert(USER_DELETED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_RMD(self, path)
+    
+    def ftp_DELE(self, path):
+        self.__send_alert(USER_DELETED_FILE_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_DELE(self, path)
+
+
+class FTPAlertingServer(FTPServer):
+    def __init__(self, *args, **kwargs):
+        self.alerting_function = kwargs.pop("alerting_function")
+        self.base_dir = kwargs.pop("base_dir")
+        FTPServer.__init__(self, *args, **kwargs)
+
+
+class FTPService(ServerCustomService):
+    def __init__(self, *args, **kwargs):
+        super(FTPService, self).__init__(*args, **kwargs)
+        self.server = None
+        self.temp_dir = None
+
+    def prepare_temp_dir(self):
+        self.temp_dir = tempfile.mkdtemp()
+
+    def delete_temp_dir(self):
+        shutil.rmtree(self.temp_dir, ignore_errors=True)
+
+    def on_server_start(self):
+        self.prepare_temp_dir()
+        authorizer = DummyAuthorizer()
+        authorizer.add_user(DEFAULT_USER, DEFAULT_PASSWORD, homedir=self.temp_dir, perm='elradfmw')  # All permissions
+        authorizer.add_anonymous(homedir=self.temp_dir, perm='elradfmw')
+
+        handler = AlertingHandler
+        handler.authorizer = authorizer
+        self.server = FTPAlertingServer(
+            (SERVER_IP, 21),
```

Comment:
Should make 21 a const too, and use it in a test() function that you should also add ;)